node/src/isolate/external.cpp

```diff
@@ -241,60 +254,75 @@ struct external_t::inner_t :
     }
 
     void connect() {
-        if(connecting) {
-            COCAINE_LOG_INFO(log, "connection to isolate daemon is already in progress");
+        if (std::chrono::system_clock::now() < (last_failed_connect_time + seal_time)) {
+            COCAINE_LOG_WARNING(log, "connection to isolate daemon is sealed");
             return;
         }
-        if(std::chrono::system_clock::now() < (last_failed_connect_time + seal_time)) {
-            COCAINE_LOG_WARNING(log, "connection to isolate daemon is sealed");
+
+        if (connecting.test_and_set()) {
+            COCAINE_LOG_INFO(log, "connection to isolate daemon is already in progress");
             return;
         }
-        connecting = true;
-        socket.reset(new asio::ip::tcp::socket(io_context));
-        if(session) {
+
+        if (session) {
             session->detach(std::error_code());
             session = nullptr;
         }
-        auto ep = args.as_object().at("external_isolation_endpoint", dynamic_t::empty_object).as_object();
+
+        const auto ep = args.as_object().at("external_isolation_endpoint", dynamic_t::empty_object).as_object();
         tcp::endpoint endpoint(
             asio::ip::address::from_string(ep.at("host", "127.0.0.1").as_string()),
             static_cast<unsigned short>(ep.at("port", 29042u).as_uint())
         );
 
-        auto connect_timeout_ms = args.as_object().at("connect_timeout_ms", 5000u).as_uint();
+        const auto connect_timeout_ms = args.as_object().at("connect_timeout_ms", 5000u).as_uint();
 
         COCAINE_LOG_INFO(log, "connecting to external isolation daemon to {}", boost::lexical_cast<std::string>(endpoint));
+
         auto self_shared = shared_from_this();
+        auto sentry = std::make_shared<flag_sentry_t>(connecting);
+        socket = std::make_unique<tcp::socket>(io_context);
 
         connect_timer.expires_from_now(boost::posix_time::milliseconds(connect_timeout_ms));
-        connect_timer.async_wait([=](const std::error_code& ec){
-            if(!ec) {
-                self_shared->socket->cancel();
+        connect_timer.async_wait([=, sentry_=sentry](const std::error_code& ec) {
+            // Async handler could be called from another thread.
+            if (ec) {
+                return;
+            }
+
+            COCAINE_LOG_INFO(log, "cancelling connection to isolate daemon");
+            // Just close any last created socket.
+            if (socket) {
```

Comment:
Why can't it be just as simple as:
* extend this check with `if (socket && socket.get() == expected_socked)` where `expected_socket` is address captured
* replace `socket->cancel();` by `socket->close();` and check in `async_connect` callback that `socket` is still open
* keep previous logic behind `connecting` - only `async_connect` should clear it, but store it as `std::atomic_flag`

Reply:
Reasonable schema, even check `if (socket && socket.get == expected_socket)` could be simplified to `if (socket.get() == expected_socket)`, as `expected_socket` can't be null.

But it seems that there is still a race as socket check for expected and socket->close is not atomic operation. In our case it is not critical as all async. callbacks and `inner_t::connect` operations (except first one from `external_t` constructor) are executed in one thread, but code could be changed someday.

And `connecting` sentry in timer callback serves as kind of global lock on any currently active socket operation, making `inner_t::connect` thread safe, so whatever async. operation will complete last will release the lock.